# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [11]:
def addTwo(x):
    return x + 2

print(addTwo(4))

#Lambda Function Syntax
#example of calling a lambda function without a variable
print((lambda x: x + 2)(4))


6
6


#### Saving to a Variable

In [5]:
add_two = lambda x: x +2
add_two(4)


6

#### Multiple Inputs

In [12]:
#multiple inputs with no variable
print((lambda x,y,z: x * y *z)(3,5,8))

#multiple inputs with a variable
multiple_values = lambda x,y,z: x + y + z

print(multiple_values(3,5,8))

120
16


#### Passing a Lambda into a Function

In [14]:
def multiply(f, num):
    """
    f expects to be a lambda function
    num expects a number
    """
    return f(num)

# when called, lambda will appear in 1st position, number will appear in 2nd

multiply(lambda x: x * x, 4) # lambda 4: 4 * 4


16

#### Returning a Lambda from a Function

In [25]:
# Regular defined function:
def multiply_test(num):
    return num * 4

#print(multiply_test(5))

# function within a function (loosely labeled as a decorator)
# DEFINTELY needs REVIEW
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# lambda function from a regular function
def returnLamb(b,c):
    return lambda x,a: x + a + b + c

r_lamb = returnLam(4,5)
print(r_lamb(5,5))

# regular function that returns a function with variables
def return_Lamb(b,c):
    def returnedValue(x,a):
        return x + a + b + c
    return returnedValue

regular_lamb = return_Lamb(4,5)

print(regular_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x0000011D670E39D0>
8
19
19


#### If Statements within Lambdas

In [32]:
# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [31]:
def cube_number(x):
    return x * x * x

print(cube_number(5))
print(cube_number(10))
print(cube_number(20))


print((lambda x: x * x *x)(5))
print((lambda x: x * x *x)(10))
print((lambda x: x * x *x)(20))


f = lambda cube: cube ** 3
    
print(f(5))
print(f(10))
print(f(20))





125
1000
8000
125
1000
8000
125
1000
8000


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [48]:
#typically used with lists. Could be used with a dictionary, tuple, or set
#begins with the initial iteration of your iterable
#Normally use the map function

def squared (num1,num2):
    if num1 < 10 and num2 < 10:
        return num1 **2, num2 ** 2
    else:
        return num1, num2

numbers = [4,11,20,3,15]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared,numbers,more_nums))
print(squared_nums_map)

#try to replicate with multiple loops

#def squared(num1, num2):

for i in numbers:
    print(i)
    #if num1 < 10:
      #  print(i **2) #return num1 ** 2
    #else:
        #print(i) #return num1
        #for j in numbers:
         #   if num2 < 10:
          #      return num2 ** 2
           # else:
            #    return num2
        
#print(square(numbers,more_nums))



[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]
4
11
20
3
15


#### Using Lambda's with Map

In [52]:
# start with the norma; syntax map(Lambda x: x **2, iterable...)

square_nums_lamb = list(map(lambda x,y: (x ** 2, y **2) if x < 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)

NameError: name 'squared_nums_lamb' is not defined

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [59]:
listx = [1,12,3,5,7]
listy = [21,3,7,22,4]

squared_minus = list(map(lambda x,y: (x *2 -1, y * 2 - 1) if x > 0 and y > 0 else (x,y), listx, listy))
print(squared_minus)


listx = [1,12,3,5,7]


squared_minus = list(map(lambda x: (x *2 -1) if x > 0 else (x), listx,))
print(squared_minus)






[(1, 41), (23, 5), (5, 13), (9, 43), (13, 7)]
[1, 23, 5, 9, 13]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [220]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica', " "]

def a_names(name):
    if name[1].lower() =='o':
        return True
    else:
        return False

new_names_list = list(filter(a_names, names))
print(new_names_list)


IndexError: string index out of range

#### Using Lambda's with Filter()

In [102]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == "a" else False))
print(new_names_list)

TypeError: filter expected 2 arguments, got 1

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [134]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]

print(mean(nums))

avg = mean(nums)
print(avg)

def mean_test(name):
    if name < avg:
        return True
    else:
        return False
new_calc = list(filter(mean_test,nums))
print(new_calc)

# Lambda function solution
below_avg = list(filter(lambda x: True if x <= mean(nums) else False, nums))
print(below_avg)

4.728571428571429
4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [132]:
from functools import reduce

# reduce( function, iterable)

list_1 = [2,4,6,8,10]

def addnums(num1,num2):
    return num1 + num2

result_add = reduce(addnums, list_1)
print(result_add)




3840


#### Using Lambda's with Reduce()

In [131]:
result_lamb = reduce(lambda x,y: x + y, list_1)
print(result_lamb)



30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [133]:
#24 is the answer

my_list = [1,2,3,4]

#def mult_nums(num1,num2):
#    return num1 * num2


result_mult = reduce(lambda x,y: x * y, my_list)
print(result_mult)



24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [141]:
def addNums(num):
    # Set base case for recusive function
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        
        return num + addNums(num - 1)
    
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [161]:
# 5! = 5 * 4 * 3 * 2 * 1

def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num-1)

factorial(20)





2432902008176640000

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [174]:
def subtract_nums(num):
    # Set base case for recusive function
    if num <= 1:
        print("subtract_nums(1) = 1")
        return num
    else:
        print(f"subtract_nums({num}) = {num} - subtract_nums({num - 1})")
        return num - subtract_nums(num - 1)
    
print(subtract_nums(5))


subtract_nums(5) = 5 - subtract_nums(4)
subtract_nums(4) = 4 - subtract_nums(3)
subtract_nums(3) = 3 - subtract_nums(2)
subtract_nums(2) = 2 - subtract_nums(1)
subtract_nums(1) = 1
3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [188]:
def my_range(stop,start,step = 2):
    while start < stop:
        yield start #yield keyword denotes a generator
        start += step

image_list = []

for i in my_range(20, start =2):
    my_generator_value = i
    print(my_generator_value)
    image_list.append(my_generator_value)

#yield allows you to go through all of your values

print(image_list)

2
4
6
8
10
12
14
16
18
[2, 4, 6, 8, 10, 12, 14, 16, 18]


#### Infinite Generator

In [ ]:
# bad, never create infinite loop

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [194]:
def yield_squared(stop, start, step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in yield_squared(0, start = 30):
    print(i)






900
841
784
729
676
625
576
529
484
441
400
361
324
289
256
225
196
169
144
121
100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [269]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

new_names_lamb = list(filter(lambda name: name.strip(), places))
print(new_names_lamb)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [270]:
# sort by the last name of the author
#sort by the final index in each respective string

author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda name: name.split (" ") [-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [274]:
# F = (9/5)*C + 32
place = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# I solved the problem in two ways.  The first is more efficient, but I'll post the second one too
#because it took a lot of time

#More efficient



places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
my_cities = list(map(lambda x: (x[0]), places))
my_conversion = list(map(lambda x: (x[-1] * 1.8 + 32), places))
my_output = list(zip(my_cities, my_conversion))
print(my_output)


#Less efficient

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
temp_list = []
city_list = []
for i in places:
   temp_list.append(i[-1])
   city_list.append(i[0])
conversion = list(map(lambda x: x * 1.8 +32, temp_list))
output = list(zip(city_list, conversion))
print(output)








                 

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [272]:
#this is the 6th iteration, since 0 is used as the base

def fibonacci(n):
    #basecase is n == 0
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        print(f"fibonacci({n}) = {n-1} + fibonacci({n - 2})")
        return fibonacci(n-1) + fibonacci(n-2)
    
print(fibonacci(6))

fibonacci(6) = 5 + fibonacci(4)
fibonacci(5) = 4 + fibonacci(3)
fibonacci(4) = 3 + fibonacci(2)
fibonacci(3) = 2 + fibonacci(1)
fibonacci(2) = 1 + fibonacci(0)
fibonacci(2) = 1 + fibonacci(0)
fibonacci(3) = 2 + fibonacci(1)
fibonacci(2) = 1 + fibonacci(0)
fibonacci(4) = 3 + fibonacci(2)
fibonacci(3) = 2 + fibonacci(1)
fibonacci(2) = 1 + fibonacci(0)
fibonacci(2) = 1 + fibonacci(0)
8
